[JIT] Support returning tuple from custom bound C++ method

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32555 [JIT] Serialize attributes and types in ClassType serialization
* #32553 [JIT] Support for registering C++ lambdas as methods on custom C++ class
* #32514 Revert "Remove __torch__ from custom class qualname"
* #32508 [JIT] Add torch.classes.load_library
* **#32477 [JIT] Support returning tuple from custom bound C++ method**
* #32471 [JIT] Fix custom class method binding for const methods
* #32470 [JIT] Test __getstate__ and __setstate__ for custom bound C++ classes
* #32260 [JIT] Passing custom class as arg

Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

Commit:
[JIT] Support returning tuple from custom bound C++ method

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit d98ab81b:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d98ab81b74ccfad4bcb8e7d332600c77a3d3d879).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4312770) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89199246) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89199246))
<details>
<summary>
<code>Jan 22 01:35:43 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:67:12: error: chosen constructor is explicit in copy-initialization</code>
</summary>

```
Jan 22 01:35:41 [ 93%] Building CXX object test_api/CMakeFiles/test_api.dir/misc.cpp.o 
Jan 22 01:35:41 [ 93%] Building CXX object test_jit/CMakeFiles/test_jit.dir/test_module_api.cpp.o 
Jan 22 01:35:42 Scanning dependencies of target shm 
Jan 22 01:35:42 [ 93%] Building CXX object caffe2/torch/lib/libshm/CMakeFiles/shm.dir/core.cpp.o 
Jan 22 01:35:42 [ 93%] Building CXX object test_api/CMakeFiles/test_api.dir/module.cpp.o 
Jan 22 01:35:42 [ 93%] Linking CXX shared library ../../../../lib/libshm.so 
Jan 22 01:35:43 [ 93%] Building CXX object test_api/CMakeFiles/test_api.dir/modulelist.cpp.o 
Jan 22 01:35:43 [ 93%] Built target shm 
Jan 22 01:35:43 [ 93%] Building CXX object test_api/CMakeFiles/test_api.dir/modules.cpp.o 
Jan 22 01:35:43 [ 93%] Building CXX object test_jit/CMakeFiles/test_jit.dir/test_interpreter.cpp.o 
Jan 22 01:35:43 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:67:12: error: chosen constructor is explicit in copy-initialization 
Jan 22 01:35:43     return {1337.0f, 123}; 
Jan 22 01:35:43            ^~~~~~~~~~~~~~ 
Jan 22 01:35:43 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:612:19: note: explicit constructor declared here 
Jan 22 01:35:43         constexpr tuple(_U1&& __a1, _U2&& __a2) 
Jan 22 01:35:43                   ^ 
Jan 22 01:35:43 In file included from /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:1: 
Jan 22 01:35:43 /var/lib/jenkins/workspace/torch/custom_class.h:117:36: warning: 'op' is deprecated: Registering operator kernels with stateful lambdas (i.e. lambdas with a capture) has non-obvious behavior. This is deprecated. Please use a lambda without a capture or a functor class instead. [-Wdeprecated-declarations] 
Jan 22 01:35:43         torch::RegisterOperators().op(qualFuncName, std::move(func))); 
Jan 22 01:35:43                                    ^ 
Jan 22 01:35:43 /var/lib/jenkins/workspace/torch/custom_class.h:153:5: note: in instantiation of function template specialization 'torch::jit::class_<torch::jit::(anonymous namespace)::Foo>::defineMethod<long, (lambda at /var/lib/jenkins/workspace/torch/custom_class.h:150:17)>' requested here 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4312764) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89199279) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89199279))
<details>
<summary>
<code>Jan 22 01:36:00 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:67:25: error: converting to 'std::tuple<double, long int>' from initializer list would use explicit constructor 'constexpr std::tuple<_T1, _T2>::tuple(_U1&&, _U2&&) [with _U1 = float; _U2 = int; <template-parameter-2-3> = void; _T1 = double; _T2 = long int]'</code>
</summary>

```
Jan 22 01:35:58 [ 90%] Linking CXX shared library ../../../../lib/libshm.so 
Jan 22 01:35:58 [ 90%] Built target shm 
Jan 22 01:35:59 Scanning dependencies of target caffe2_detectron_ops 
Jan 22 01:35:59 [ 90%] Building CXX object test_api/CMakeFiles/test_api.dir/nn_utils.cpp.o 
Jan 22 01:35:59 [ 90%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/ps_roi_pool_op.cc.o 
Jan 22 01:35:59 [ 90%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/spatial_narrow_as_op.cc.o 
Jan 22 01:35:59 [ 90%] Building CXX object test_api/CMakeFiles/test_api.dir/optim.cpp.o 
Jan 22 01:36:00 [ 90%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/upsample_nearest_op.cc.o 
Jan 22 01:36:00 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp: In instantiation of 'std::tuple<double, long int> torch::jit::{anonymous}::Stack<T>::return_a_tuple() const [with T = std::__cxx11::basic_string<char>]': 
Jan 22 01:36:00 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:88:53:   required from here 
Jan 22 01:36:00 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:67:25: error: converting to 'std::tuple<double, long int>' from initializer list would use explicit constructor 'constexpr std::tuple<_T1, _T2>::tuple(_U1&&, _U2&&) [with _U1 = float; _U2 = int; <template-parameter-2-3> = void; _T1 = double; _T2 = long int]' 
Jan 22 01:36:00      return {1337.0f, 123}; 
Jan 22 01:36:00                          ^ 
Jan 22 01:36:00 In file included from /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:1:0: 
::__cxx11::string, Func) [with R = long int; Func = torch::jit::class_<CurClass>::def_(std::__cxx11::string, Func, torch::jit::detail::types<R, Types ...>) [with Func = long int (torch::jit::{anonymous}::Foo::*)(); R = long int; Types = {}; CurClass = torch::jit::{anonymous}::Foo; std::__cxx11::string = std::__cxx11::basic_string<char>]::<lambda(c10::intrusive_ptr<torch::jit::{anonymous}::Foo>)>; CurClass = torch::jit::{anonymous}::Foo; std::__cxx11::string = std::__cxx11::basic_string<char>]': 
Jan 22 01:36:00 /var/lib/jenkins/workspace/torch/custom_class.h:153:20:   required from 'torch::jit::class_<CurClass>& torch::jit::class_<CurClass>::def_(std::__cxx11::string, Func, torch::jit::detail::types<R, Types ...>) [with Func = long int (torch::jit::{anonymous}::Foo::*)(); R = long int; Types = {}; CurClass = torch::jit::{anonymous}::Foo; std::__cxx11::string = std::__cxx11::basic_string<char>]' 
Jan 22 01:36:00 /var/lib/jenkins/workspace/torch/custom_class.h:106:20:   required from 'torch::jit::class_<CurClass>& torch::jit::class_<CurClass>::def(std::__cxx11::string, Func) [with Func = long int (torch::jit::{anonymous}::Foo::*)(); CurClass = torch::jit::{anonymous}::Foo; std::__cxx11::string = std::__cxx11::basic_string<char>]' 
Jan 22 01:36:00 /var/lib/jenkins/workspace/test/cpp/jit/test_custom_class.cpp:74:47:   required from here 
:is_functor<LambdaType>::value && (! c10::guts::detail::is_stateless_lambda_<typename std::decay<typename std::decay<_Tp>::type>::type, void>::value)), c10::RegisterOperators&&> = c10::RegisterOperators&&; std::__cxx11::string = std::__cxx11::basic_string<char>]' is deprecated: Registering operator kernels with stateful lambdas (i.e. lambdas with a capture) has non-obvious behavior. This is deprecated. Please use a lambda without a capture or a functor class instead. [-Wdeprecated-declarations] 
Jan 22 01:36:00      registeredOps().push_back( 
Jan 22 01:36:00      ^ 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Support returning tuple from custom bound C++ method"


Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support returning tuple from custom bound C++ method"


Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Great catch!

Commit:
Update on "[JIT] Support returning tuple from custom bound C++ method"


Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support returning tuple from custom bound C++ method"


Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support returning tuple from custom bound C++ method"


Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support returning tuple from custom bound C++ method"


Differential Revision: [D19509927](https://our.internmc.facebook.com/intern/diff/D19509927)

[ghstack-poisoned]
